ocaml/networkd/lib/network_device_order.ml

```diff
@@ -0,0 +1,526 @@
+
+  let parse output_of_one_dev =
+    debug "%s: line: %s" __FUNCTION__ output_of_one_dev ;
+    let kvs =
+      let open Astring.String in
+      cuts ~sep:"\n" output_of_one_dev
+      |> List.filter_map (fun line ->
+             cut ~sep:":" line |> Option.map (fun (k, v) -> (trim k, trim v))
+         )
+    in
+    List.iter (fun (k, v) -> debug "%s: [%s]=[%s]" __FUNCTION__ k v) kvs ;
+    [
+      ( "BIOS device"
+      , fun acc value -> {acc with bios_eth_order= n_of_ethn value}
+      )
+    ; ("Kernel name", fun acc value -> {acc with name= value})
+    ; ( "Assigned MAC"
+      , fun acc value -> {acc with mac= Macaddr.of_string_exn value}
+      )
+    ; ("Bus Info", fun acc value -> {acc with pci= Pciaddr.of_string_exn value})
+    ]
+    |> List.fold_left
+         (fun (missing_keys, acc) (key, f) ->
+           match List.assoc_opt key kvs with
+           | Some v ->
+               (missing_keys, f acc v)
+           | None ->
+               (key :: missing_keys, acc)
+         )
+         ([], default)
+    |> fun (missing_keys, r) ->
+    match missing_keys with
+    | [] ->
+        debug "%s: %s" __FUNCTION__ (to_string r) ;
+        r
+    | key :: _ ->
+        raise (Missing_key key)
+
+  exception Duplicate_mac_address
+
+  let update_multinic currents =
+    let pci_cnt =
+      let f o = Some (Option.fold ~none:1 ~some:(fun c -> c + 1) o) in
+      List.fold_left
+        (fun acc dev -> PciaddrMap.update dev.pci f acc)
+        PciaddrMap.empty currents
+    in
+    List.map
+      (fun dev : t ->
+        let multinic =
+          (* Will never raise exception or be < 1 *)
+          let c = PciaddrMap.find dev.pci pci_cnt in
+          if c > 1 then true else false
+        in
+        {dev with multinic}
+      )
+      currents
+
+  let get_all () : t list =
+    try
+      let all =
+        Network_utils.call_script cmd_biosdevname
+          ["--policy"; "all_ethN"; "-d"; "-x"]
+        |> Astring.String.cuts ~sep:"\n\n"
+        |> List.filter (fun line -> line <> "")
+        |> List.map parse
+        |> update_multinic
+      in
+      ListToMacaddrMap.to_11_map ~by:(fun v -> v.mac) all |> ignore ;
+      all
+    with
+    | ListToMacaddrMap.Duplicate_key ->
+        raise Duplicate_mac_address
+    | e ->
+        error "%s" "Can't parse the output of the biosdevname!" ;
+        raise e
+end
+
+(** A network device which has been in an order. *)
+module OrderedDev = struct
+  type t = Network_interface.ordered_iface
+
+  let compare_on_mac t1 t2 = Macaddr.compare t1.mac t2.mac
+
+  let to_string t =
+    Printf.sprintf "position=%d; name=%s; MAC=%s; PCI=%s; present=%s" t.position
+      t.name (Macaddr.to_string t.mac) (Pciaddr.to_string t.pci)
+      (string_of_bool t.present)
+
+  let map_by_pci (l : t list) : t list PciaddrMap.t =
+    ListToPciaddrMap.to_1n_map ~by:(fun v -> v.pci) l
+
+  let map_by_position (l : t list) : t IntMap.t =
+    ListToIntMap.to_11_map ~by:(fun v -> v.position) l
```

Comment:
This doesn't transform the Duplicate_key exception, and neither of its users in the module handle it.

This exception should not exit the module and should be handled in some way.

Reply:
All errors is changed to use `Result`.